### ISSUE TITLE
Inclusion of in-tree driver?

### ISSUE THREAD

**Issue by jarkkojs** [keystone-enclave/keystone#372]
Out-of-tree drivers are used in embedded projects only if boilerplate shim is needed to adapt some proprietary asset. Otherwise it is inferior choice for having

1. A single Linux kernel source tree.
2. A single unified kernel image, which has all "power on" features.

It does not really matter if the project is purely academic research or a commercial project because this the most efficient model.

I've written a driver that is API/ABI compatible with the out-of-tree driver. It is available at:

https://git.kernel.org/pub/scm/linux/kernel/git/jarkko/linux-riscv-keystone.git/

See `keystone-5.19.y` branch.

I'm proposing this to be included to the main project as a patch file.

**Comment by jarkkojs** [comment]
There is one difference in ABI, i.e. `library_name` is passed as pointer instead of string array.

**Comment by jarkkojs** [comment]
> There is one difference in ABI, i.e. `library_name` is passed as pointer instead of string array.

apparently has been from upstream good

**Comment by jarkkojs** [comment]
I don't want necessarily own the code but I could hand-over the patch for you and whatever you have in mind can build on top. I just put it there so that you can look into it. I.e. I'm not looking into maintaining this. Just wanted clear that out.

The motivation is debugging SM and SDK. If everything working all is cool but the out-of-tree driver is not great with rollback. With my driver things keep online even if stuff fails around. I've experienced total systems freezes with it because of unorthodox use of buddy allocator.

To summarize is not kernel per se but instead of working on the Keystone SDK, and I happy to hand over this to you. I read that you are making some changes to the enclave boot-up model. I still think that this driver code base provides a good for memory management so I would consider making that change on top of this.

**Comment by jarkkojs** [comment]
I decided that I intend to keep my alternative driver in-sync and update it to be API compatible whatever gets changed in the out-of-tree driver for the time being. It is clean and stable code base to maintain in its current condition.
